### PR TITLE
[Mime][BrowserKit] Add FormUrlEncoded - Fix content-type in HttpBrowserPost

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30867 
| License       | MIT
| Doc PR        | NA

This PR use the content-type `x-www-form-urlencoded` when posting a form with the HttpBrowser.

### CODE REVIEW

File: src/Symfony/Component/Mime/Part/UrlEncodedPart.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use Symfony\Component\Mime\Encoder\QpContentEncoder;
+use Symfony\Component\Mime\Exception\InvalidArgumentException;
+use Symfony\Component\Mime\Header\Headers;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ *
+ * @experimental in 4.3
+ */
+final class UrlEncodedPart extends AbstractPart
```
Comment (jderusse) [inline src/Symfony/Component/Mime/Part/UrlEncodedPart.php]: Should not be a Part. I'm removing it.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
@jderusse Can you apply the patch from fabbot?

**Comment by fabpot** [comment]
... and we also need some tests ofc :)

File: src/Symfony/Component/BrowserKit/HttpBrowser.php [hunk 2]
```diff
@@ -43,44 +43,52 @@ public function __construct(HttpClientInterface $client = null, History $history
...
             'body' => $body,
             'max_redirects' => 0,
         ]);
 
         return new Response($response->getContent(false), $response->getStatusCode(), $response->getHeaders(false));
     }
 
-    private function getBody(Request $request): ?AbstractPart
+    /**
+     * @return array contain body and headers
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/HttpBrowser.php]: ` @return array [$body, $headers]` maybe?

File: src/Symfony/Component/BrowserKit/HttpBrowser.php [hunk 3]
```diff
@@ -43,44 +43,52 @@ public function __construct(HttpClientInterface $client = null, History $history
...
             $fields[$name] = DataPart::fromPath($file['tmp_name'], $file['name']);
         }
 
-        return new FormDataPart($fields);
+        if ($hasFile) {
+            $part = new FormDataPart($fields);
+            return [$part->bodyToIterable(), $part->getPreparedHeaders()->toArray()];
+        }
+
+        return [http_build_query($fields, '', '&', PHP_QUERY_RFC1738), ['Content-Type' => 'x-www-form-urlencoded']];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/HttpBrowser.php]: it's `application/x-www-form-urlencoded` right?

File: src/Symfony/Component/BrowserKit/Tests/HttpBrowserTest.php [hunk 4]
```diff
@@ -112,4 +114,62 @@ public function testGetInternalResponse()
...
+                $this->assertInstanceOf('\Generator', $options['body']);
+                $this->assertContains('my_file', (\implode('', \iterator_to_array($options['body']))));
+
+                return true;
+            }))
+            ->willReturn($this->createMock(ResponseInterface::class));
+
+        $browser = new HttpBrowser($client);
+        $path = \tempnam(\sys_get_temp_dir(), 'http');
+        \file_put_contents($path, 'my_file');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/Tests/HttpBrowserTest.php]: no `\`

**Review by nicolas-grekas — APPROVED** [review]
(with minor comments)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jderusse.

**Pull request merged**